### PR TITLE
Added TraceReporter to trace AsyncWriteJournal, SnapshotStore and Serializer operations.

### PR DESCRIPTION
**Pull request by s10myk4**
TraceReporter provides interfaces to retrieve traces of AsyncWriteJournal, SnapshotStore and Serializer operations.

### CODE REVIEW

File: journal/src/main/scala/com/github/j5ik2o/akka/persistence/dynamodb/journal/DynamoDBJournal.scala [hunk 1]
```diff
@@ -355,38 +363,42 @@ class DynamoDBJournal(config: Config) extends AsyncWriteJournal with ActorLoggin
...
-    val context                       = MetricsReporter.newContext(UUID.randomUUID(), pid)
+    val context                       = Context.newContext(UUID.randomUUID(), pid)
     val newContext                    = metricsReporter.fold(context)(_.beforeJournalAsyncReadHighestSequenceNr(context))
+
     def fetchHighestSeqNr(): Future[Long] = {
       journalDao
         .highestSequenceNr(PersistenceId.apply(persistenceId), SequenceNumber(fromSequenceNr)).runWith(Sink.head)
     }
+
     val future = writeInProgress.get(persistenceId) match {
```
Comment (j5ik2o) [inline journal/src/main/scala/com/github/j5ik2o/akka/persistence/dynamodb/journal/DynamoDBJournal.scala:387]: Change to def.

File: journal/src/main/scala/com/github/j5ik2o/akka/persistence/dynamodb/serialization/ByteArrayJournalSerializer.scala [hunk 2]
```diff
@@ -75,7 +77,7 @@ class ByteArrayJournalSerializer(
       index: Option[Int]
   )(implicit ec: ExecutionContext): Future[JournalRow] = {
     val pid        = PersistenceId(persistentRepr.persistenceId)
-    val context    = MetricsReporter.newContext(UUID.randomUUID(), pid)
+    val context    = Context.newContext(UUID.randomUUID(), pid)
     val newContext = metricsReporter.fold(context)(_.beforeJournalSerializeJournal(context))
 
     val future = for {
```
Comment (j5ik2o) [inline journal/src/main/scala/com/github/j5ik2o/akka/persistence/dynamodb/serialization/ByteArrayJournalSerializer.scala:83]: Change to def.

File: journal/src/main/scala/com/github/j5ik2o/akka/persistence/dynamodb/serialization/ByteArrayJournalSerializer.scala [hunk 3]
```diff
@@ -89,37 +91,38 @@ class ByteArrayJournalSerializer(
...
 
   override def deserialize(
       journalRow: JournalRow
   )(implicit ec: ExecutionContext): Future[(PersistentRepr, Set[String], Long)] = {
     val pid        = journalRow.persistenceId
-    val context    = MetricsReporter.newContext(UUID.randomUUID(), pid)
+    val context    = Context.newContext(UUID.randomUUID(), pid)
     val newContext = metricsReporter.fold(context)(_.beforeJournalDeserializeJournal(context))
 
     val future = for {
```
Comment (j5ik2o) [inline journal/src/main/scala/com/github/j5ik2o/akka/persistence/dynamodb/serialization/ByteArrayJournalSerializer.scala:112]: Change to def.

File: journal/src/main/scala/com/github/j5ik2o/akka/persistence/dynamodb/journal/DynamoDBJournal.scala [hunk 4]
```diff
@@ -341,7 +349,7 @@ class DynamoDBJournal(config: Config) extends AsyncWriteJournal with ActorLoggin
     implicit val ec: ExecutionContext = defaultExecutionContext
 
     val pid        = PersistenceId(persistenceId)
-    val context    = MetricsReporter.newContext(UUID.randomUUID(), pid)
+    val context    = Context.newContext(UUID.randomUUID(), pid)
     val newContext = metricsReporter.fold(context)(_.beforeJournalAsyncReplayMessages(context))
     val future = journalDao
```
Comment (j5ik2o) [inline journal/src/main/scala/com/github/j5ik2o/akka/persistence/dynamodb/journal/DynamoDBJournal.scala:354]: Change to def.

**Review by j5ik2o — CHANGES_REQUESTED** [review]

File: journal/src/main/scala/com/github/j5ik2o/akka/persistence/dynamodb/journal/DynamoDBJournal.scala [hunk 5]
```diff
@@ -318,21 +325,22 @@ class DynamoDBJournal(config: Config) extends AsyncWriteJournal with ActorLoggin
 
   override def asyncDeleteMessagesTo(persistenceId: String, toSequenceNr: Long): Future[Unit] = {
     val pid        = PersistenceId(persistenceId)
-    val context    = MetricsReporter.newContext(UUID.randomUUID(), pid)
+    val context    = Context.newContext(UUID.randomUUID(), pid)
     val newContext = metricsReporter.fold(context)(_.beforeJournalAsyncDeleteMessagesTo(context))
 
     implicit val ec: ExecutionContext = defaultExecutionContext
 
     val future = journalDao
```
[the remark this answers is not in the document]
Comment (s10myk4) [reply journal/src/main/scala/com/github/j5ik2o/akka/persistence/dynamodb/journal/DynamoDBJournal.scala]: fixed.
https://github.com/j5ik2o/akka-persistence-dynamodb/pull/943/commits/91e2814a4fcf841b337569bf886e735fd5f6d93e#diff-761788fe0403c57a933e6fcf54d2517b298b1a0afcc9e4ef3586ac5b03c3289eR355

File: journal/src/main/scala/com/github/j5ik2o/akka/persistence/dynamodb/journal/DynamoDBJournal.scala (hunk 1 above)
Comment (s10myk4) [reply journal/src/main/scala/com/github/j5ik2o/akka/persistence/dynamodb/journal/DynamoDBJournal.scala]: fixed.
https://github.com/j5ik2o/akka-persistence-dynamodb/pull/943/commits/91e2814a4fcf841b337569bf886e735fd5f6d93e#diff-761788fe0403c57a933e6fcf54d2517b298b1a0afcc9e4ef3586ac5b03c3289eR388

File: journal/src/main/scala/com/github/j5ik2o/akka/persistence/dynamodb/journal/DynamoDBJournal.scala [hunk 6]
```diff
@@ -411,20 +423,21 @@ class DynamoDBJournal(config: Config) extends AsyncWriteJournal with ActorLoggin
...
 
     val pid        = PersistenceId(persistenceId)
-    val context    = MetricsReporter.newContext(UUID.randomUUID(), pid)
+    val context    = Context.newContext(UUID.randomUUID(), pid)
     val newContext = metricsReporter.fold(context)(_.beforeJournalAsyncUpdateEvent(context))
     val write      = PersistentRepr(message, sequenceNumber, persistenceId)
 
     serializer
       .serialize(write).flatMap { serializedRow =>
         val future = journalDao.updateMessage(serializedRow).runWith(Sink.ignore)
```
[the remark this answers is not in the document]
Comment (s10myk4) [reply journal/src/main/scala/com/github/j5ik2o/akka/persistence/dynamodb/journal/DynamoDBJournal.scala]: fixed.
https://github.com/j5ik2o/akka-persistence-dynamodb/pull/943/commits/91e2814a4fcf841b337569bf886e735fd5f6d93e#diff-761788fe0403c57a933e6fcf54d2517b298b1a0afcc9e4ef3586ac5b03c3289eR433

File: journal/src/main/scala/com/github/j5ik2o/akka/persistence/dynamodb/journal/DynamoDBJournal.scala (hunk 4 above)
Comment (s10myk4) [reply journal/src/main/scala/com/github/j5ik2o/akka/persistence/dynamodb/journal/DynamoDBJournal.scala]: fixed.
https://github.com/j5ik2o/akka-persistence-dynamodb/pull/943/commits/91e2814a4fcf841b337569bf886e735fd5f6d93e#diff-761788fe0403c57a933e6fcf54d2517b298b1a0afcc9e4ef3586ac5b03c3289eR355

File: journal/src/main/scala/com/github/j5ik2o/akka/persistence/dynamodb/serialization/ByteArrayJournalSerializer.scala (hunk 2 above)
Comment (s10myk4) [reply journal/src/main/scala/com/github/j5ik2o/akka/persistence/dynamodb/serialization/ByteArrayJournalSerializer.scala]: fixed.
https://github.com/j5ik2o/akka-persistence-dynamodb/pull/943/commits/91e2814a4fcf841b337569bf886e735fd5f6d93e#diff-599e2fb130750a3c53ab98d064e4eddec7688c71eb9548e4074946a727aace74R83

File: journal/src/main/scala/com/github/j5ik2o/akka/persistence/dynamodb/serialization/ByteArrayJournalSerializer.scala (hunk 3 above)
Comment (s10myk4) [reply journal/src/main/scala/com/github/j5ik2o/akka/persistence/dynamodb/serialization/ByteArrayJournalSerializer.scala]: fixed. 
https://github.com/j5ik2o/akka-persistence-dynamodb/pull/943/commits/91e2814a4fcf841b337569bf886e735fd5f6d93e#diff-599e2fb130750a3c53ab98d064e4eddec7688c71eb9548e4074946a727aace74R112

**Pull request merged**